File: Canon/Fullrepupdate/iskra_mainworkline_updatedSot/tools/verify_ledger.py
#!/usr/bin/env python3
"""Verify ledger/sot.json hashes."""
from __future__ import annotations
import hashlib, json, os, sys
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]

def sha256_file(path: Path) -> str:
    h = hashlib.sha256()
    with path.open("rb") as f:
        for chunk in iter(lambda: f.read(1024*1024), b""):
            h.update(chunk)
    return h.hexdigest()

def main() -> None:
    ledger = ROOT / "ledger" / "sot.json"
    if not ledger.exists():
        print("Missing ledger/sot.json", file=sys.stderr)
        sys.exit(2)
    sot = json.loads(ledger.read_text(encoding="utf-8"))
    bad = []
    for rel, expected in sot.get("sha256", {}).items():
        path = ROOT / rel
        if not path.exists():
            bad.append((rel, "missing", expected))
            continue
        got = sha256_file(path)
        if got != expected:
            bad.append((rel, got, expected))
    if bad:
        print("Ledger verification FAILED:")
        for rel, got, exp in bad[:50]:
            print(f"- {rel}: got {got} expected {exp}")
        sys.exit(1)
    print(f"Ledger OK ({len(sot.get('sha256', {}))} files)")
    sys.exit(0)

if __name__ == "__main__":
    main()

File: tools/verify_ledger.ts
import * as fs from 'fs';
import * as path from 'path';
import * as crypto from 'crypto';

const ROOT = process.cwd();

function sha256_file(filePath: string): string {
    const fileBuffer = fs.readFileSync(filePath);
    const hashSum = crypto.createHash('sha256');
    hashSum.update(fileBuffer);
    return hashSum.digest('hex');
}

function main() {
    const ledgerPath = path.join(ROOT, 'ledger', 'sot.json');
    if (!fs.existsSync(ledgerPath)) {
        console.error('Missing ledger/sot.json');
        process.exit(2);
    }
    
    const sot = JSON.parse(fs.readFileSync(ledgerPath, 'utf-8'));
    const bad: [string, string, string][] = [];
    const entries = sot.sha256 || {};
    
    for (const [rel, expected] of Object.entries(entries)) {
        const filePath = path.join(ROOT, rel);
        if (!fs.existsSync(filePath)) {
            bad.push([rel, 'missing', expected as string]);
            continue;
        }
        
        const got = sha256_file(filePath);
        if (got !== expected) {
            bad.push([rel, got, expected as string]);
        }
    }
    
    if (bad.length > 0) {
        console.error('Ledger verification FAILED:');
        for (const [rel, got, exp] of bad.slice(0, 50)) {
            console.error(`- ${rel}: got ${got} expected ${exp}`);
        }
        process.exit(1);
    }
    
    console.log(`Ledger OK (${Object.keys(entries).length} files)`);
    process.exit(0);
}

main();
